python/ray/train/data_parallel_trainer.py

```diff
@@ -254,25 +252,66 @@ def __init__(
         train_loop_config: Optional[Dict] = None,
         backend_config: Optional[BackendConfig] = None,
         scaling_config: Optional[ScalingConfig] = None,
-        dataset_config: Optional[Dict[str, DatasetConfig]] = None,
+        dataset_config: Optional[DataConfig] = None,
         run_config: Optional[RunConfig] = None,
         datasets: Optional[Dict[str, GenDataset]] = None,
-        preprocessor: Optional["Preprocessor"] = None,
         resume_from_checkpoint: Optional[Checkpoint] = None,
+        # Deprecated.
+        preprocessor: Optional["Preprocessor"] = None,
     ):
         self._train_loop_per_worker = train_loop_per_worker
         self._train_loop_config = train_loop_config
 
+        if isinstance(dataset_config, dict) or self._dataset_config or preprocessor:
+            # Warn about deprecated cases (will raise error in future).
+            if isinstance(dataset_config, dict):
+                logger.warning(
+                    "The dict form of `dataset_config` is deprecated. Use the "
+                    "DataConfig class instead. Support for this will be dropped "
+                    "in a future release."
+                )
+            elif preprocessor:
+                logger.warning(
+                    "The `preprocessor` argument to Trainer is deprecated. Apply "
+                    "dataset transformations ahead of time instead. Support for this "
+                    "will be dropped in a future release."
+                )
+            # If using the new API, hard-disallow deprecated features.
+            if isinstance(dataset_config, DataConfig):
+                if self._dataset_config:
+                    raise ValueError(
+                        "The new DataConfig class is not supported for use with the "
+                        "legacy Trainer._dataset_config field."
```

Comment:
can you add information on what should be done to avoid this error?

In this case, this should only affect people creating their own Trainer subclasses.